include/kernels/dslash_wilson.cuh

```diff
@@ -121,16 +120,16 @@ namespace quda
           Link U = arg.U.Ghost(d, gauge_ghost_idx, 1 - gauge_parity);
           HalfVector in = arg.in.Ghost(d, 0, ghost_idx + coord.s * arg.dc.ghostFaceCB[d], their_spinor_parity);
 
-          out += (conj(U) * in).reconstruct(d, proj_dir);
+          out += (conj(U) * in).template reconstruct<d,proj_dir>();
         } else if (doBulk<kernel_type>() && !ghost) {
 
           Link U = arg.U(d, gauge_idx, 1 - gauge_parity);
           Vector in = arg.in(back_idx + coord.s * arg.dc.volume_4d_cb, their_spinor_parity);
 
-          out += (conj(U) * in.project(d, proj_dir)).reconstruct(d, proj_dir);
+          out += (conj(U) * in.template project<d, proj_dir>()).template reconstruct<d,proj_dir>();
         }
       }
-    } // nDim
+    }); // nDim
```

Comment:
Do you know if there is an easy way to fall back to `pragma unroll for` (easy meaning having a macro define or something like that)? Seems to make the templates work we need to add this additional `);` to the end of each for loop, and there is no easy way to turn the template variables into C++ variables easily.

Reply:
Blame C++ for obnoxious syntax.  I can't figure out a way to deal with the ending of the block, apart from defining a pair of macros like `BEGIN_STATIC_FOR`/`END_STATIC_FOR`.

The lambda in the C++17 version of the `static_for` just uses `auto`.  Instead of template parameters, the functions in `color_spinor.h` could also be defined with `auto` arguments, though the function body would be more complicated depending on how much we want to help the compiler in optimizing the switch statements.